src/os/bluestore/bluestore_types.h

```diff
@@ -272,6 +272,121 @@ static inline bool operator!=(const bluestore_extent_ref_map_t& l,
   return !(l == r);
 }
 
+/// blob_use_tracker: a set of per-alloc unit ref counters to track blob usage
+struct bluestore_blob_use_tracker_t {
+  // N.B.: There is no need to minimize ref_size/refs_count
+  //   as much as possible (e.g. have just a single byte for ref_size) since:
+  //   1) Struct isn't packed hence it's padded. And even if it's packed see 2)
+  //   2) Mem manager has its own granularity, most probably >= 8 bytes
+  //
+  uint32_t ref_size;
+  uint32_t refs_count; 
+                       
+  union {
+    uint32_t* ref_by_au;
+    uint32_t ref_total;
+  };
+  
+  bluestore_blob_use_tracker_t()
+    : ref_size(0), refs_count(0), ref_by_au(nullptr) {
+  }
+  ~bluestore_blob_use_tracker_t() {
+    clear();
+  }
+
+  void clear() {
+    if (refs_count != 0) {
+      delete ref_by_au;
+    }
+    ref_by_au = 0;
+    ref_size = 0;
+  }
+
+  uint32_t get_referenced() const {
+    uint32_t total = ref_total;
+    if (refs_count) {
+      total = 0;
+      for (size_t i = 0; i < refs_count; ++i) {
+	total += ref_by_au[i];
+      }
+    }
+    return total;
+  }
+  bool is_not_empty() const {
+    if (refs_count) {
+      for (size_t i = 0; i < refs_count; ++i) {
+	if (ref_by_au[i]) {
+	  return true;
+	}
+      }
+      return false;
+    }
+    return ref_total != 0;
+  }
+
+  void get(
+    uint32_t offset,
+    uint32_t len,
+    std::function<void(uint32_t*, uint32_t*)> fallback_cb);
+
+  bool put(
+    uint32_t offset,
+    uint32_t len,
+    std::function<void(uint32_t*, uint32_t*)> fallback_cb,
+    set<uint32_t> *release_units);
+
+  void split(
+    uint32_t blob_offset,
+    bluestore_blob_use_tracker_t* r);
+
+  bool equal(
+    const bluestore_blob_use_tracker_t& other) const;
+    
+  void bound_encode(size_t& p) const {
+    denc(refs_count, p);
```

Comment:
Does that mean that we have a bug in bluestore_extent_ref_map_t::bound_encode?
` void bound_encode(size_t& p) const {`
`    denc((uint32_t)0, p);`
`   ...`